### ISSUE TITLE
Question about Dataset

### ISSUE THREAD

**Issue by kjwoo31** [ai4r/AIR-Act2Act.pytorch#4]
<img src="https://user-images.githubusercontent.com/59794238/90993388-a831c580-e5ef-11ea-8134-617c234e6c70.png" width="800" height="500"> (P001-P050, Train 과정에 대한 Confusion Matrix)   

Dataset에 대해 확인한 결과, 'threaten to hit with right hand'에 대한 데이터가 포함되지 않은 것 같습니다. 확인 부탁드립니다.

**Comment by kjwoo31** [comment]
Train Result
![image](https://user-images.githubusercontent.com/59794238/91004790-21411500-e610-11ea-8b06-fb5b733ceab6.png) ![image](https://user-images.githubusercontent.com/59794238/91004759-11293580-e610-11ea-9733-5c930e1e663a.png)
![image](https://user-images.githubusercontent.com/59794238/91004762-14242600-e610-11ea-9781-4172fc9b0f9f.png) ![image](https://user-images.githubusercontent.com/59794238/91004767-171f1680-e610-11ea-80d6-61aa76f58c7a.png)

 Train하였을 때, Epoch에 따라 Training Loss가 증가하고 Training Accuracy가 감소하는 결과가 발생하였는데 이에 대해서도 확인 부탁드립니다.

**Comment by wrko** [comment]
recog_subaction.py -m test 로 테스트 해보니

‘A008’ 데이터에 대한 인식 결과를 보면

12번 행위 (threaten to hit with right hand)가 잘 생성되네요.

 



 

다시 한번 확인해 보시거나,

아니면 confusion matrix를 출력해 본 코드를 공유해 주시면

확인해보도록 하겠습니다.

 

고우리 드림.

**Comment by wrko** [comment]
recog_subaction.py -m test 로 테스트 해보니
‘A008’ 데이터에 대한 인식 결과를 보면
12번 행위 (threaten to hit with right hand)가 잘 생성되네요.

![test](https://user-images.githubusercontent.com/13827622/91010583-b696d600-e61d-11ea-9115-1fc290741a95.png)

다시 한번 확인해 보시거나,
아니면 confusion matrix를 출력해 본 코드를 공유해 주시면
확인해보도록 하겠습니다.

**Comment by kjwoo31** [comment]
confusion matrix는 recog_subaction.py에 아래의 코드를 추가하여 출력하였습니다. 

                for i in valid_outputs:
                    for j in valid_predictions:
                        conf_matrix[i][j] += 1
----------------------------

    labels = ["stand", "open the door", "hand on wall", "not shown", "raise right hand", "wave right hand", "lower right hand"
                , "cry with right hand", "raise both hands", "cry with both hands", "lower left hand or both hands", "raise or cry with left hand"
                , "threaten to hit with right hand",  "raise or threaten to hit with left hand"]
    plt.figure(figsize=(20, 20))
    sns.heatmap(conf_matrix, annot=True, fmt='.0f', xticklabels=labels, yticklabels=labels, linewidths=.5)
    plt.title('confusion matrix')
    plt.show()


![image](https://user-images.githubusercontent.com/59794238/91012104-7dac3080-e620-11ea-8fd3-1a70bbe048a3.png)

**Comment by kjwoo31** [comment]
확인결과, Preprecess 과정에서 데이터가 누락되면서 문제가 발생한 것 같습니다. Preprocess.py를 다시 실행하였더니 아래와 같이 12번 행위 (threaten to hit with right hand)가 생성되는 것을 확인하였습니다.
![image](https://user-images.githubusercontent.com/59794238/91015381-08435e80-e626-11ea-8eeb-08a7387b9378.png)

**Comment by wrko** [comment]
아 그런가요??
Preprocess 과정에서 데이터는 왜 누락이 됐었던 건가요??

**Comment by kjwoo31** [comment]
죄송합니다. 제가 착각했습니다. 
 Train data를 사용하여 confusion matrix를 구성하여야 하는데, validation data를 사용하여 confusion matrix를 만들어 데이터가 누락되었었습니다.
(Preprocess 과정에서 전체 data 중 일부에 대해 validation data로 저장하는데, 우연히 12번 행위 (threaten to hit with right hand)만 누락된 것 같습니다.)
 다시 마지막 Epoch의 Train data에 대해 Confusion Matrix를 그린 후 comment를 게시하고 issue를 close 하겠습니다.

**Comment by wrko** [comment]
아 그렇군요. 궁금해서 물어봤습니다.
혹시 더 문제가 있으면 알려주세요.

**Comment by kjwoo31** [comment]
<img width="1915" alt="제목 없음" src="https://user-images.githubusercontent.com/59794238/91106941-070a4400-e6af-11ea-8730-8e56442dd6f3.png">

 confusion matrix가 위 사진과 같이 나타났습니다. 이에 대해 2가지 질문이 있습니다.
 1. 아래 그래프와 같이 학습을 진행할수록 loss가 늘고 accuracy가 줄어드는데 Epoch 수가 모자라서 생기는 문제인가요?
 2. Repository에서 제공되는 model_0013.pth은 어떻게 학습하셨나요?

> Train Result
> ![image](https://user-images.githubusercontent.com/59794238/91004790-21411500-e610-11ea-8b06-fb5b733ceab6.png) ![image](https://user-images.githubusercontent.com/59794238/91004759-11293580-e610-11ea-9733-5c930e1e663a.png)
> ![image](https://user-images.githubusercontent.com/59794238/91004762-14242600-e610-11ea-9781-4172fc9b0f9f.png) ![image](https://user-images.githubusercontent.com/59794238/91004767-171f1680-e610-11ea-80d6-61aa76f58c7a.png)
> 
> Train하였을 때, Epoch에 따라 Training Loss가 증가하고 Training Accuracy가 감소하는 결과가 발생하였는데 이에 대해서도 확인 부탁드립니다.

(confusion matrix code)

    conf_matrix = np.zeros([14,14])
-----------------------------------------
            if epoch == num_epochs:
                for i in train_outputs:
                    for j in train_predictions:
                        conf_matrix[i][j] += 1
-----------------------------------------
    for i in range(0, 14):
        n_sum = sum(conf_matrix[i])
        for j in range(0, 14):
            conf_matrix[i][j] = conf_matrix[i][j] / n_sum
--------------------------------------------
    labels = ["stand", "open the door", "hand on wall", "not shown", "raise right hand", "wave right hand", "lower right hand"
                , "cry with right hand", "raise both hands", "cry with both hands", "lower left hand or both hands", "raise or cry with left hand"
                , "threaten to hit with right hand",  "raise or threaten to hit with left hand"]
    plt.figure(figsize=(20, 20))
    sns.heatmap(conf_matrix, annot=True, fmt='.0%', xticklabels=labels, yticklabels=labels, linewidths=.5)
    plt.title('confusion matrix')
    plt.show()

**Comment by wrko** [comment]
The reason the confusion matrix values ​​are weird is that the K-means clustering models were loaded in a different sklearn version than the trained environment.

Please change the version of sklearn to 0.22.1.
$ pip uninstall scikit-learn
$ pip install scikit-learn==0.22.1

**Comment by kjwoo31** [comment]
scikit-learn의 버전을 바꿔 학습한 결과, 다음과 같이 confusion matrix가 나타났습니다. 정상적으로 train된 게 맞는지 확인 부탁드립니다.

<img width="1915" alt="200" src="https://user-images.githubusercontent.com/59794238/91248273-43ad6c80-e78f-11ea-820e-03bba59cb84c.png">

![image](https://user-images.githubusercontent.com/59794238/91248208-195baf00-e78f-11ea-9c09-33c9a2395a62.png)

**Comment by wrko** [comment]
To draw a confusion matrix, you should change the code from
```
for i in train_outputs:
    for j in train_predictions:
        conf_matrix[i][j] += 1
```
to
```
for i, j in zip(train_outputs, train_predictions):
    conf_matrix[i][j] += 1
```

In addition, I added the code to check accuracy, loss, and confusion matrix of a trained model.
Run ```python recog_subaction.py -m validate``` and enter the model number.
You can check the confusion matrix on a pop-up window.

**Comment by kjwoo31** [comment]
![image](https://user-images.githubusercontent.com/59794238/91508707-3a4d0d00-e913-11ea-8ead-1b9e3e8a10d0.png)
Validation Loss: 0.12736, Validation Acc: 0.97270

확인하였습니다. 감사합니다.

**Comment by wrko** [comment]
네 고생 많았어요!
